### PR TITLE
[BREAKING] Switch to torchtext

### PR DESCRIPTION
**Pull request by srush**
This PR switches the system to utilize torchtext instead of our own data management. This is a relatively big change, so I am looking for tests to try it out on. 

* Removes entirely: `Loss.py` `Dataset.py`, and `Dict.py`, major reduction in code to `translate.py` and `preprocess.py`
* Introduces new dependencies on `torchtext` and `dill`

This is the final major breaking change to pyopennmt I hope to make this summer. 

Tests:

* run my suite of models: translation, summarization, im2text ... 
* run sebastian's test

### CODE REVIEW

**Comment by jekbradbury** [comment]
Wondering if you guys might be in a good place to make use of https://github.com/pytorch/text -- there's a very old branch refactored to use it [here](https://github.com/MetaMind/OpenNMT-py/tree/torchtext) -- or if not perhaps you have thoughts about what we're missing? We'd love to get torchtext to the point where it can satisfy the use cases of real libraries like OpenNMT, but we're not doing ourselves any favors dragging our feet on test/docs/PyPI...

**Comment by srush** [comment]
Sure, I think we would love to use it! I wasn't sure if it was ready yet? 

There are a couple features which OpenNMT would need. Are these in there yet?

- Support for word features. We have been using datasets like "word|feat1|feat2 word|feat1|feat2..."
- Support for length vectors aligned with batches.

**Comment by jekbradbury** [comment]
The second is supported using `include_lengths=True` for a Field. The easiest way to add word features would probably be to use an additional Field for each kind of feature, and add custom logic in the Dataset constructor that passes each word to a featurizer (the same way that the SNLI dataset constructor passes each sentence to a parse converter and stores the result in a new Field).

**Comment by srush** [comment]
Okay, I'll give it a try. I would love to get rid of this code.

**Comment by srush** [comment]
So far this is great. Just curious if there is a canonical way to attach vector data that is not numericalized. For instance can I have a field that is an image where it is a tensor ?

**Comment by jekbradbury** [comment]
If you create a Field with `use_vocab=False` it should skip numericalization and allow you to pass raw tensors, but I'm not sure if it'll exactly do what you're looking for -- if not that should be pretty easy to add.

**Comment by srush** [comment]
Okay, so far it's good and way less code. Couple more questions:

- Any way to serialize datasets or fields or vocab? At minimum we need vocab for test. Torch.save seems to fail. For long datasets it's also convenient to create dicts before training.

- Had to write my own iterator for in batch sorting. Needed for rnn pack

- use_vocab works well. Wrote a custom tensor_type to do the merge.

- Assuming we cant control indexing of fields at all?  That's probably better, our code hackily assume padding is 0.

**Comment by jekbradbury** [comment]
1. Currently serialization with `pickle` is broken because of lambdas as attributes. But `torch.save` should still work if you pass `dill` as the optional pickle module argument; if that’s not an acceptable dependency the alternatives are for us to clean all lambdas out of torchtext or to provide a `state_dict` property like we already do on `Iterator`s.
2. Examples within each batch are sorted by default (in `BucketIterator` as well as the standard `Iterator` with `sort=True`) according to the provided `sort_key`. However, they’re sorted in ascending rather than descending order, so you’d have to negate the provided key in order to have an order that’s accepted by the PyTorch RNNs. The problem is that `sort_key` is also what’s used to assign similar-length examples to similar batches. In your case the most efficient choice for the key would be interlaced source and target lengths, to minimize padding, but in order to get sorting within batches you’d have to sort by the source sentence length alone, which is maybe 5-10 percent more wasteful on decoder padding. Or you could keep your custom iterator that gets you both the ideal sort between batches and the correct one within them; you might want to confirm that the second sort isn’t a significant overhead (it shouldn’t be if you use torch.sort).
3. Lots of crazy duck typing opportunities were intentionally included in torchtext but I don’t think I anticipated that one!
4. Torchtext uses 0 as unk for all vocabs basically to avoid special casing and allow every vocab to be exactly a Python defaultdict. If every vocab is a defaultdict then every vocab has an unk, but since only sequential fields have padding and I wanted indices to be as consistent as possible unk takes 0 and padding gets 1. Honestly that’s not a very good reason, and I’m open to allowing additional flexibility if it would be helpful.

Also, something that would help the padding and efficiency situation but would also be a more fundamental change for OpenNMT: you could make use of the support for dynamic batch sizes. By providing a lambda to `batch_size_fn` you can allow batches of up to some max number of total tokens rather than a max number of examples; this can give a speed-up of up to 2x on datasets with lots of variation in sentence length and is employed by e.g. the tensor2tensor library.

**Comment by srush** [comment]
Oh I like the idea of using `batch_size_fn`. I'll add that. Also `dill` mostly worked for me, still needed to add a `__getstate__` to my dataset because of a `generator` somewhere in the internals which `dill` can't handle. 

Now I am hacking through the translator code. Curious if there is any inverse to `numericalize` that handles removing padding etc. ? 

This is a net removal of 500 lines of code and several files. Think this is a major improvement though to outsource this from the codebase.

**Comment by jekbradbury** [comment]
We don’t currently have any helper functions for inverting numericalize etc. but I agree that something like that would be helpful. Ideally the whole process would be reversible, but that would require at least a fully reversible tokenizer (probably sentencepiece-style) which I haven’t gotten around to adding.
